BuildingBlocks/test/BuildingBlocks.Application.Tests/Mediatr/QuotaEnforcerBehaviorTests.cs

```diff
@@ -1,87 +1,194 @@
+    }
+
+    [Fact]
+    public async void Succeeds_when_the_passed_quota_is_not_exhausted()
+    {
+        // Arrange
+        var behavior = CreateQuotaEnforcerBehavior(TestData.MetricStatus.ThatIsNotExhausted);
+        var nextHasRan = false;
+
+        // Act
+        Func<Task> acting = async () => await behavior.Handle(
+            new TestCommand(),
+            () =>
+            {
+                nextHasRan = true;
+                return Task.FromResult(new IResponse());
+            },
+            CancellationToken.None);
+
+        await Task.Run(acting, CancellationToken.None);
+
+        // Assert
+        nextHasRan.Should().BeTrue();
+    }
+
+    [Fact]
+    public async void Succeeds_when_the_passed_quota_was_exhausted()
+    {
+        // Arrange
+        var behavior = CreateQuotaEnforcerBehavior(TestData.MetricStatus.ThatWasExhaustedUntilYesterday);
+        var nextHasRan = false;
+
+        // Act
+        Func<Task> acting = async () => await behavior.Handle(
+            new TestCommand(),
+            () =>
+            {
+                nextHasRan = true;
+                return Task.FromResult(new IResponse());
+            },
+            CancellationToken.None);
+
+        await Task.Run(acting, CancellationToken.None);
+
+        // Assert
+        nextHasRan.Should().BeTrue();
+    }
 
-    public QuotaEnforcerBehaviorTests() {
-        _userContextStub = new UserContextStub();
+    [Fact]
+    public void Throws_QuotaExhaustedException_when_the_passed_quota_is_still_exhausted()
+    {
+        // Arrange
+        var behavior = CreateQuotaEnforcerBehavior(TestData.MetricStatus.ThatIsExhaustedFor10Days);
+
+        // Act
+        Func<Task> acting = async () => await behavior.Handle(
+            new TestCommand(),
+            () =>
+            {
+                return Task.FromResult(new IResponse());
+            },
+            CancellationToken.None);
+
+        // Assert
+        acting.Should().AwaitThrowAsync<QuotaExhaustedException>()
+            .Which.ExhaustedMetricStatuses.First().MetricKey.Should().Be(TestData.MetricStatus.ThatIsExhaustedFor10Days.MetricKey);
     }
 
-    [ApplyQuotasForMetrics("*")]
-    private class TestCommand : IRequest { }
+    /// <summary>
+    /// This test uses a special implementation of the IMetricStatusesRepository which <strong>
+    /// returns zero results</strong> on the `GetMetricStatuses` Method.
+    /// For this reason, the MetricStatuses passed to the constructor have zero relevance for the test.
+    /// </summary>
+    [Fact]
+    public async void Succeeds_when_none_of_the_metrics_exist_in_the_repository()
+    {
+        // Arrange
+        var metricStatusesStubRepository = new MetricStatusesNoMatchStubRepository(new() {
+            TestData.MetricStatus.ThatWasExhaustedUntilYesterday, TestData.MetricStatus.ThatIsNotExhausted
+        });
```

Comment:
Wtf, that's the ugliest piece of code I've ever seen. You pass something to a constructor, which has to exist because it is needed by the base class even though the base class doesn't give the sub class anything relevant. How would anyone understand this code without digging deeper? 😲